Workflowv2 workflows create

garbage-collector opened the pull request:
Prior to starting a PR, please make sure you have read our
[contributor tutorial](https://github.com/gophercloud/gophercloud/tree/master/docs/contributor-tutorial).

Prior to a PR being reviewed, there needs to be a Github issue that the PR
addresses. Replace the brackets and text below with that issue number.

For #1199 

https://github.com/openstack/mistral/blob/master/mistral/api/controllers/v2/workflow.py#L141

garbage-collector commented:
Ready to review!

jtopjian commented on openstack/workflow/v2/workflows/requests.go:
```diff
@@ -0,0 +1,56 @@
+		}
+		url += query
+		b = tmpB
+	}
```
Can this instead be:

```go
if opts != nil {
  query, err := opts.ToWorkflowCreatQuery()
  if err != nil {
    r.Err = err
    return
  }
  url += query
  b = opts.Definition
}
```

Or is there a reason to have `opts.Definition` come from `ToWorkflowCreateQuery()`?

garbage-collector replied:
I based this from what I saw in objectstorage package, because workflows have the same way of working regarding the create process: the POST request takes a definition file as body (text/plain), and other parameters are sent through URL query.

This is how it's handled on objectstorage:
https://github.com/gophercloud/gophercloud/blob/master/openstack/objectstorage/v1/objects/requests.go#L171 

I just kept the same signature. I'll probably rename the function "ToWorkflowCreateQuery" to "ToWorkflowCreateParams" to be less ambiguous.

We can't make your proposal as we have a CreateOptsBuilder which is an interface, so no access to opts.Definition.

jtopjian replied:
> I based this from what I saw in objectstorage package, because workflows have the same way of working regarding the create process: the POST request takes a definition file as body (text/plain), and other parameters are sent through URL query.

Right, and re-using existing patterns is definitely the way to go. At least as a start.

What I was pointing out is that you aren't actually doing anything with `Definition` inside `ToWorkflowCreateParams`, though. For example, in the `objects` package, the `Content` is used to create the etag hash.

My question was more _will_ you ever be doing anything with `Definition`? 

However,

> We can't make your proposal as we have a CreateOptsBuilder which is an interface, so no access to opts.Definition.

Oh, right - duh. I totally overlooked that. Busy day :)

jtopjian replied:
> What do you think about it?

Nah, that'll be a large break from the current patterns. Let's leave it as-is.

jtopjian commented on openstack/workflow/v2/workflows/results.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// Extract helps to get a Workflow struct from a Get function.
+func (r commonResult) Extract() (*Workflow, error) {
```
If I'm understanding correctly, "create" will return a list (https://github.com/openstack/mistral/blob/master/mistral/api/controllers/v2/workflow.py#L169-L173), even if there's only one workflow, but GET will return a single workflow?

If so, then maybe have a `GetResult` and a `CreateResult` rather than one inheriting from the other - might add better separation.

Also, slight nit, but this shouldn't be in this PR. No doubt this was accidental, though, so it's OK to keep it. 😄

jtopjian commented on openstack/workflow/v2/workflows/results.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// Workflow represents a workflow execution on OpenStack mistral API.
+type Workflow struct {
```
It appears these fields are defined here:

https://github.com/openstack/mistral/blob/632e4bda67ca1e6b579607f264f258890b0c1d8e/mistral/api/controllers/v2/resources.py#L88-L101

(for future PRs, we'll need references like this to validate the returned fields and their types)

It looks like this is missing a few fields such as tags, created_at, and updated_at?

garbage-collector replied:
Added missing fields

garbage-collector commented:
Thanks for the review !

All comments have been answered and fixed, and I rebased all workflowsv2-workflows-* branches with the required fixes.

But it looks like I've made some mess with my rebases as we lost some coverage on files that I didn't touch at all.

Can you help me to understand what happened ? Not really used to git workflows like this (dependencies between branches and stuff like this), so I'd love to have some help to understand what happened and what I missed.

garbage-collector commented:
BTW @jtopjian :
I've ran the various tests and they were successfull. I can provide you an access to a devstack working with Mistral if you want, just give me your pubkey and I can spawn a devstack to perform the acceptance tests. Not an issue to build one :)

jtopjian commented:
> But it looks like I've made some mess with my rebases as we lost some coverage on files that I didn't touch at all.

Don't worry about the test coverage results. It can be quite buggy at times. 

As well, you don't need to worry about the OpenLab tests, either, since Mistral isn't set up. Any OpenLab failures are probably transient. 

This is why running the acceptance tests offline is important for this series of PRs.

> I can provide you an access to a devstack working with Mistral if you want

Thank you for the offer! I should be able to get Mistral set up within the next week or so - just pressed for time at the moment :)

jtopjian commented on openstack/workflow/v2/workflows/results.go:
```diff
@@ -53,4 +42,13 @@ type Workflow struct {
+	Tags []string `json:"tags"`
+
+	// CreatedAt is the creation date of the workflow.
+	CreatedAt gophercloud.JSONRFC3339ZNoTNoZ `json:"created_at"`
```
We'll want this to be of type `time.Time` and you'll need to implement a custom `UnmarshalJSON` func. See here for an example: https://github.com/gophercloud/gophercloud/blob/master/docs/FAQ.md#overriding-default-unmarshaljson-method

Or just grep/search for `UnmarshalJSON` in this repository as this happens frequently.

jtopjian commented on openstack/workflow/v2/workflows/results.go:
```diff
@@ -53,4 +42,13 @@ type Workflow struct {
+	CreatedAt gophercloud.JSONRFC3339ZNoTNoZ `json:"created_at"`
+
+	// UpdatedAt is the last update date of the workflow.
+	UpdatedAt *gophercloud.JSONRFC3339ZNoTNoZ `json:"updated_at"`
```
Same here

garbage-collector commented:
Fixed Created and Updated field to time.Time.

I rebased all the workflows PR with this.

Thanks :)
